core_lib/src/managers/backupmanager.cpp

```diff
@@ -0,0 +1,711 @@
+/*
+
+Pencil2D - Traditional Animation Software
+Copyright (C) 2005-2007 Patrick Corrieri & Pascal Naidon
+Copyright (C) 2008-2009 Mj Mendoza IV
+Copyright (C) 2012-2020 Matthew Chiawen Chang
+
+This program is free software; you can redistribute it and/or
+modify it under the terms of the GNU General Public License
+as published by the Free Software Foundation; version 2 of the License.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+*/
+
+#include "object.h"
+#include "editor.h"
+
+#include <QDebug>
+#include <QSettings>
+
+
+#include "preferencemanager.h"
+#include "layermanager.h"
+#include "soundmanager.h"
+#include "backupmanager.h"
+#include "viewmanager.h"
+#include "selectionmanager.h"
+#include "soundmanager.h"
+
+#include "backupelement.h"
+#include "legacybackupelement.h"
+
+#include "layerbitmap.h"
+#include "layercamera.h"
+#include "layervector.h"
+#include "layersound.h"
+
+
+#include "bitmapimage.h"
+#include "vectorimage.h"
+#include "soundclip.h"
+#include "camera.h"
+
+BackupManager::BackupManager(Editor* editor) : BaseManager(editor, "BackupManager")
+{
+    qDebug() << "BackupManager: created";
+}
+
+BackupManager::~BackupManager()
+{
+    qDebug() << "BackupManager: destroyed";
+}
+
+bool BackupManager::init()
+{
+    mUndoStack = new QUndoStack(this);
+    qDebug() << "BackupManager: init";
+
+    mNewBackupSystemEnabled = editor()->preference()->isOn(SETTING::NEW_UNDO_REDO_SYSTEM_ON);
+
+    return true;
+}
+
+Status BackupManager::load(Object* /*o*/)
+{
+    clearStack();
+    return Status::OK;
+}
+
+Status BackupManager::save(Object* /*o*/)
+{
+    if (newUndoRedoSystemEnabled()) {
+        mBackupAtSave = latestBackupElement();
+    } else {
+        mLegacyBackupAtSave = mLegacyBackupList[mLegacyBackupIndex];
+    }
+    return Status::OK;
+}
+
+void BackupManager::backup(BackupType type)
+{
+    if (!newUndoRedoSystemEnabled()) {
+        return;
+    }
+
+    Layer* currentLayer = editor()->layers()->currentLayer();
+    switch (type)
```

Comment:
Fair point, although I initially decided to not have the description be an argument to avoid that the tools might do something like this themselves. I also didn't want to enable that you could potentially write an undo description which wouldn't have to be related to the UndoRedoType argument type. In hindsight maybe that's a bit silly.

I'll address it 👍

Reply:
I've simplified the function but I'm reluctant to make the description available when there's also the UndoRedoType. Isn't it better to encapsulate the localizable strings here rather than let the caller do whatever or do do you have concerns in terms of the potential complexity of the switch?

Reply:
Changes: 

add -> record(...)
saveStates -> state(...)

I've reworked how the undo state is created and how it's given to the record function.
The UndoRedoType -> UndoRedoRecordType is now incorporated into the state function and control what kind of state it returns. 

The description is as requested now an argument again to the record method, so every tool uses the typeName() description for the undo/redo action.

layerType from undoState is now used instead of current layer.

I think that should address all your concerns 🤔